compiler/parser/RdrHsSyn.hs

```diff
@@ -2236,18 +2237,71 @@ data ChoiceConsuming = PreConsuming | Consuming | PostConsuming | NonConsuming
   deriving Show -- for creating type/data names from these constructors
 
 data FlexChoiceData = FlexChoiceData {
-    fcdControllers :: LHsExpr GhcPs
-  , fcdObservers   :: Maybe (LHsExpr GhcPs)
-  , fcdChoiceData  :: ChoiceData
+    fcdChoiceBodyDecls :: Located [Located ChoiceBodyDecl]
+  , fcdChoiceData    :: ChoiceData
   }
 
+data ChoiceParties = ChoiceParties {
+    cpControllers :: LHsExpr GhcPs
+  , cpObservers   :: Maybe (LHsExpr GhcPs)
+  , cpAuthorizers :: Maybe (LHsExpr GhcPs)
+  }
+
+data ChoiceBodyDecl
+  = ChoiceControllerDecl (LHsExpr GhcPs)
+  | ChoiceObserverDecl (LHsExpr GhcPs)
+  | ChoiceAuthorityDecl (LHsExpr GhcPs)
+
+data ChoiceBodyDecls = ChoiceBodyDecls {
+      cbdControllers :: [LHsExpr GhcPs]
+    , cbdObservers :: [LHsExpr GhcPs]
+    , cbdAuthrizers :: [LHsExpr GhcPs]
+    }
+
+instance Semigroup ChoiceBodyDecls where
+  (<>) x y = ChoiceBodyDecls {
+      cbdControllers = cbdControllers x Monoid.<> cbdControllers y
+    , cbdObservers = cbdObservers x Monoid.<> cbdObservers y
+    , cbdAuthrizers = cbdAuthrizers x Monoid.<> cbdAuthrizers y
+    }
+
+instance Monoid ChoiceBodyDecls where
+  mempty = ChoiceBodyDecls [] [] []
+
+choiceBodyDeclToDecls :: Located ChoiceBodyDecl -> ChoiceBodyDecls
+choiceBodyDeclToDecls (L _ decl) = case decl of
+  ChoiceControllerDecl x -> mempty { cbdControllers = [x] }
+  ChoiceObserverDecl x -> mempty { cbdObservers = [x] }
+  ChoiceAuthorityDecl x -> mempty { cbdAuthrizers = [x] }
+
+extractChoiceBodyDecls :: [Located ChoiceBodyDecl] -> ChoiceBodyDecls
+extractChoiceBodyDecls = foldMap choiceBodyDeclToDecls
+
+validateChoiceBodyDecls :: Located RdrName -> Located [Located ChoiceBodyDecl] -> P ChoiceParties
+validateChoiceBodyDecls choiceName (L _ decls)
+  | null cbdControllers = report "Missing 'controller' declaration"
+  | length cbdControllers > 1 = report "Multiple 'controller' declarations"
+  | length cbdObservers > 1 = report "Multiple 'observer' declarations"
+  | length cbdAuthrizers > 1 = report "Multiple 'authority' declarations"
+  | otherwise =
+    do
+      let cpControllers = head cbdControllers
+      let cpObservers = listToMaybe cbdObservers
+      let cpAuthorizers = listToMaybe cbdAuthrizers
+      pure $ ChoiceParties{..}
+  where
+    ChoiceBodyDecls{..} = extractChoiceBodyDecls decls
+    report :: String -> P a
+    report e = addFatalError (getLoc choiceName) (text e)
```

Comment:
Does it always make sense for the error to be on the name? Say I have 2 controller defs, I'd expect the error to be on the second controller

Reply:
Attaching to the choice name mirrors the approach takes for missing/multiple decls within a template, where the error attach to the template name

Reply:
But as you point out there is scope to be even more precise if we so desire.

Reply:
Perhaps a low prio QoL ticket could be made here